### ISSUE TITLE
Consider changing the vote program to only allow commission changes in the first half of an epoch

### ISSUE THREAD

**Issue by mvines** [solana-labs/solana#29346]
It's currently too easy for validator commission rugs to take place by changing commission immediately before the epoch boundary and restoring it afterwards. 

Ideally commission changes would require an epoch or more of delay before taking effect but the vote account would need to be extended to support such a delay. This is a non-trivial and time consuming task.

One simple alternative is to only permit vote account commission changes in the first half of an epoch. This gives attentive stakers and bots 24 hours to respond to unexpected commission changes, which is a significant improvement over the current situation.

Implementation-wise, this would be a low-risk change to the vote program as the existing `VoteInstruction::UpdateCommission` could be surgically modified to use the `Clock` and `EpochSchedule` sysvars to reject commission changes in the second half of an epoch.

**Comment by aeyakovenko** [comment]
What if max increment is max(1, 2*previous)

**Comment by joncinque** [comment]
> What if max increase is max(1, 2*current)

As discussed offline, someone can just spam the transaction to increase their fee if we do that, so the approach given in the PR will be best in my opinion

**Comment by aeyakovenko** [comment]
Are there any free bits in the structure?

**Comment by aeyakovenko** [comment]
We can add a pda for the warmup state.

**Comment by joncinque** [comment]
> We can add a pda for the warmup state.

That's a pretty good option too, and doesn't require any changes to the vote account. There's some gotchas in opening / closing PDAs that we have to be wary of, but other than that, it's certainly an option.

We'd need to create two new instructions: `UpdateFutureCommission` and `ApplyNewCommission` or something, which requires that PDA and some SOL to create / delete the PDA. Anyway, I've also put up a PR for this proposal

**Comment by janlegner** [comment]
> Are there any free bits in the structure?

No, as per response in the issue I opened earlier on the same topic. https://github.com/solana-labs/solana/issues/28628#issuecomment-1293889143

**Comment by t-nelson** [comment]
i'll oppose the pda path.  we're already planning to expand the vote state to make space for a couple other features.  adding the complexity of extra instruction and double account loads doesn't really make sense in that light

i'm a little dubious of the efficacy of the commission change restriction. inattentive stake is inattentive stake.  if it hasn't moved after not getting rewards, it's not going to move in the half-epoch where commission is locked.  at best it'll dissuade some cheaters who think this makes it easier to get caught.  in reality, we already have a detection method that's better than sampling and the services tracking commission changes are working towards implementing.

that said, i won't oppose the commission change restriction, but we should be very clear about setting expectations.  it wouldn't surprise me to see some cheaters just let it ride and see how much stake actually moves

**Comment by aeyakovenko** [comment]
Any docs or issues to the state changes to votes?  I am actually generally apposed to changing existing data structures unless it's a security issue. The deterministic account doesn't require a state hard fork. State hardforks are operationally expensive and hard to rollback.

**Comment by anoushk1234** [comment]
Hey @mvines do we have plans to add the epoch long delay for the commission change in the vote account? I'd love to work on it

**Comment by mvines** [comment]
> Hey @mvines do we have plans to add the epoch long delay for the commission change in the vote account? I'd love to work on it

We need to first solidify the method to grown the vote account data length (tracking issue unknown to me, if one even currently exists). This will be needed for a couple new vote account feature requests, but once that part is done then hopefully adding new functionality that requires additional state becomes less daunting.

**Comment by anoushk1234** [comment]
> > Hey @mvines do we have plans to add the epoch long delay for the commission change in the vote account? I'd love to work on it
> 
> We need to first solidify the method to grow the vote account data length (tracking issue unknown to me, if one even currently exists). This will be needed for a couple new vote account feature requests, but once that part is done then hopefully adding new functionality that requires additional state becomes less daunting.

Makes sense, I think we can open a new issue to keep track of what features need the increased vote account length and accordingly decide on the structure, once that's done we can implement each of them by priority.

**Comment by aeyakovenko** [comment]
> We need to first solidify the method to grow the vote account data length

@mvines don't we have a Realloc api that can grow the data buffer.  Should be easy to add a versioning scheme to the vote structure and grow it.

**Comment by t-nelson** [comment]
> @mvines don't we have a Realloc api that can grow the data buffer. Should be easy to add a versioning scheme to the vote structure and grow it.

vote state is already versioned.  we do have realloc, which is how we plan to grow the vote state. only open question is how to initiate and pay for the extra rent-exempt reserve
